### PR TITLE
BUG/ENH: handle spline interpolation methods in `interpn` and fix bug for trailing dimensions

### LINKED ISSUE #16472: handle spline interpolation methods in `interpn`
As long as `interpn` is (mostly) a convenience wrapper over RegularGridInterpolator, it may make sense for it to grow support for new `cubic`, `quintic` and `pchip` modes which were added for 1.9.0.

Currently it only supports `nearest` and `linear` modes of RGI, plus `splinef2d` which should probably kept for backwards compatibility.

### PR DESCRIPTION
**Pull request by j-bowhay**
#### Reference issue

Closes gh-16472
#### What does this implement/fix?

- Adds new cubic, quintic and pchip modes which were added for 1.9.0 to `interpn`.
- Fixes bug where the new spline methods did not work when the values have trailing dimensions
- Adds new test for trailing dimension
- Updated documentation for `RGI.__call__`
- A few general tidy ups and parameterisation of tests

### CODE REVIEW

**Comment by ev-br** [comment]
These tests should actually pass, by the looks of it. I'd try to drop into pdb on failure and poke around (`-- --pdb` IIRC, or just put a breakpoint into a failing test).

**Comment by j-bowhay** [comment]
There appears to be a bug hear related to these failures:
```
FAILED scipy/interpolate/tests/test_rgi.py::TestInterpN::test_xi_broadcast[cubic] - UnboundLocalError: local variable 'out_of_bounds' referenced before assignment
FAILED scipy/interpolate/tests/test_rgi.py::TestInterpN::test_xi_broadcast[quintic] - UnboundLocalError: local variable 'out_of_bounds' referenced before assignment
FAILED scipy/interpolate/tests/test_rgi.py::TestInterpN::test_xi_broadcast[pchip] - UnboundLocalError: local variable 'out_of_bounds' referenced before assignment
```

This is caused by this snippet here:
https://github.com/scipy/scipy/blob/a60206b29d548a6543d75ac95558d52492d3ecaf/scipy/interpolate/_rgi.py#L323-L339
If the method is not linear or nearest then out_of_bounds is never assigned. @ev-br do you want me to raise a separate issue?

**Comment by ev-br** [comment]
Ah, yes. This was me, doing what looked like an obvious easy cleanup. No need for a separate issue I'd say, can just fix it up (this same PR or a separate one) , would you be interested? Otherwise I'd do it later this week.

**Comment by j-bowhay** [comment]
Happy to tackle to this. how did you envisage fixing this, pulling the parts relating to `out_of_bounds` out of `_find_indices` into its own function?

**Comment by ev-br** [comment]
Exactly!

**Comment by j-bowhay** [comment]
@ev-br I notice the original issue doesn't mention `slinear`, is this intentional?

**Comment by ev-br** [comment]
@j-bowhay no, just an oversight. Would be nice to add it here, if only to avoid this question going forward.

**Comment by j-bowhay** [comment]
Ok so down to the last test failure.
```
FAILED scipy/interpolate/tests/test_rgi.py::TestInterpN::test_nonscalar_values[slinear] - ValueError: setting an array element with a sequence.
FAILED scipy/interpolate/tests/test_rgi.py::TestInterpN::test_nonscalar_values[cubic] - ValueError: setting an array element with a sequence.
FAILED scipy/interpolate/tests/test_rgi.py::TestInterpN::test_nonscalar_values[quintic] - ValueError: setting an array element with a sequence.
FAILED scipy/interpolate/tests/test_rgi.py::TestInterpN::test_nonscalar_values[pchip] - ValueError: setting an array element with a sequence.
```

Cause by this spinet:
https://github.com/scipy/scipy/blob/fb954f6a24c083ec77ba85e5ead985ae4ea88b47/scipy/interpolate/_rgi.py#L418-L431

The error is being raised as `result` has shape `(77,)` and `folded_values` has shape `(6,)`. In all the other tests `folded_values` has shape `(1,)

Before I dig deeper into this, I was wondering if you had any insight into this @ev-br ?

**Comment by ev-br** [comment]
Ugh. In that test, there is `assert_equal(v.shape, (7, 11, 6)`, which seems to hint that the problem is a missed flatten/reshape somewhere. 

On a side note, it's truly humbling to see how many issues you are uncovering in what looked like a straightforward extension. I guess I'm moving even further towards the `it's broken unless tested to the gills, and then some more` camp.

**Comment by j-bowhay** [comment]
Well at least when this goes in we gain a lot of test coverage for "free"

**Comment by j-bowhay** [comment]
I am not especially familiar with how the spline interpolation was written so please correct me @ev-br if I am barking up the wrong tree! However it seems like the current implementation was not intended to work with values that are arrays not scalars. For example this comment https://github.com/scipy/scipy/blob/497f330282f9da8eccb2cb9b801939abc81b3aa0/scipy/interpolate/_rgi.py#L425-L430
If the values are arrays not scalars then folded values is also going to be an array which means the shape of https://github.com/scipy/scipy/blob/497f330282f9da8eccb2cb9b801939abc81b3aa0/scipy/interpolate/_rgi.py#L418 is also incorrect.

**Comment by j-bowhay** [comment]
> I am not especially familiar with how the spline interpolation was written so please correct me @ev-br if I am barking up the wrong tree! However it seems like the current implementation was not intended to work with values that are arrays not scalars. For example this comment
> 
> https://github.com/scipy/scipy/blob/497f330282f9da8eccb2cb9b801939abc81b3aa0/scipy/interpolate/_rgi.py#L425-L430
> 
> 
> If the values are arrays not scalars then folded values is also going to be an array which means the shape of
> https://github.com/scipy/scipy/blob/497f330282f9da8eccb2cb9b801939abc81b3aa0/scipy/interpolate/_rgi.py#L418
> 
> is also incorrect.

Actually scratch this completely have made some progress

**Comment by ev-br** [comment]
> Actually scratch this completely have made some progress

Great! 
Let me know if you want me to do some debugging (was about to fire up pdb but happy to not interfere while you're on it).

**Comment by j-bowhay** [comment]
https://github.com/scipy/scipy/blob/79edba5dde01dcb7705ed37b3bf78ff4561298eb/scipy/interpolate/tests/test_rgi.py#L702-L721

This was the last failing test. I now have up to 715 passing (the latest commit I have pushed) without breaking the rest of the tests. However the second part of the test (721) is failing. I don't really understand what this second part of the test is trying to do! @ev-br do you have any ideas?

**Comment by j-bowhay** [comment]
> https://github.com/scipy/scipy/blob/79edba5dde01dcb7705ed37b3bf78ff4561298eb/scipy/interpolate/tests/test_rgi.py#L702-L721
> 
> This was the last failing test. I now have up to 715 passing (the latest commit I have pushed) without breaking the rest of the tests. However the second part of the test (721) is failing. I don't really understand what this second part of the test is trying to do! @ev-br do you have any ideas?

Not making much progress on this last bit. If you get a chance a hand with debugging would be appreciated

**Comment by ev-br** [comment]
Thanks for the reminder @j-bowhay . Will do when I'm back, will be a couple days though.

**Comment by ev-br** [comment]
@j-bowhay apologies it took me this long to circle back to it! I'm starting to look at this at last, and I think I understand the purpose of this test, if not the implementation (yet). TL;DR it lools like it's a limitation of RGI new interpolation modes (i.e., one more bug you're uncovering), and some of your fixes are suspect.

Looking at the `interpn` docsting, it is (cryptically) documented to accept values as `array_like, shape (m1, …, mn, …)` --- notice the ellipsis. The return value is documented to have a shape `xi.shape[:-1] + values.shape[ndim:]`. So what this test does, it checks that for `values` with trailing dimensions (that ellipsis), the result of calling `interpn` is the same as manually looping.
The original `linear` and `nearest` modes heavily rely on broadcasting, so apparently somewhere the axes end up being permuted -- something broadcasts along the first dimension. Then the `.transpose(1, 2, 0)` stanza permutes them into the have the broadcast axis to be last. But then --- *huge* kudos to the original author --- the test also checks the values, broadcast vs manual looping. And they do not match.

So,

- there is a real problem somewhere 
- As you notices, the RGI code is suspect in that it expects scalars where it should have expected arrays
- the test only probes a single extra dimension, there should also be a test with at least two trailing dimensions
- we need a direct test of this kind for the RGI itself, not only an indirect one through `interpn` (which is why we did not catch it in the original PRs)
- this all is a rather obscure feature I suspect, possibly even a case of premature vectorization, as it were. But we should fix it nonetheless :-).
- when we fix this, it needs to be documented better for the `RGI.__call__` docstring.

I'll keep looking into it.

File: scipy/interpolate/tests/test_rgi.py
Comment (ev-br) [inline scipy/interpolate/tests/test_rgi.py:23]: Ugh, what are these doing here? `test_readonly` should be moved to wherever other `griddata`  implementation is tested, `test_interpnd.py` or `test_ndgriddata.py`. Could you do it in a separate PR which we'll merge quickly @j-bowhay ?

Comment (j-bowhay) [reply scipy/interpolate/tests/test_rgi.py:23]: Yes these do seem out of place, will open a pr

Comment (j-bowhay) [reply scipy/interpolate/tests/test_rgi.py:23]: @ev-br the test for `NearestNDInterpolator`, `LinearNDInterpolator`, `CloughTocher2DInterpolator` are split over two files:

- `NearestNDInterpolator` is tested in `test_ndgriddata.py`
- `LinearNDInterpolator`, `CloughTocher2DInterpolator` are tested in `test_interpnd.py`

So can this test go in a common file that tests all 3 (if so which one would be best?) or does in need to go in both?

Comment (ev-br) [reply scipy/interpolate/tests/test_rgi.py:23]: Not that it matters all that much. I'd put them into test_ndgriddata.py, but if you've an opinion, expect no argument from me :-).

**Comment by ev-br** [comment]
One more obscure glitch masking one more problem in RGI, cf gh-16780. Might make sense to handle it quickly and rebase this one on top.

**Comment by ev-br** [comment]
https://github.com/j-bowhay/scipy/pull/1 seems to fix failures in current tests, can you take it over @j-bowhay ? Please do check   it though, this stuff is confusing so a second pair of eyes is of great help.

While I do realize that the scope of this PR crept quite a bit, I think it'd be best to implement suggestion from https://github.com/scipy/scipy/pull/16508#issuecomment-1205637922, specifically

* the test only probes a single extra dimension, there should also be a similar test with at least two trailing dimensions
* we need a direct test of this kind for the RGI itself, not only an indirect one through interpn (which is why we did not catch it in the original PRs)
* when we fix this, it needs to be documented better for the RGI.__call__ docstring.

Would be great if you could finish it up @j-bowhay

**Comment by j-bowhay** [comment]
@ev-br yes happy to finish this off. Thanks for the sorting the transpose issues, I thought something was up there but hadn't clocked that the premature transpose was causing the problems.

**Comment by ev-br** [comment]
Okay, git conflicts block the CI. Let me know if you want a hand resolving them. Now or once you add stuff already in your pipeline.

**Comment by j-bowhay** [comment]
Ok so I think the remaining stuff to be done here is:

- [ ] Test for more that one trailing dimension
- [ ] Better documentation for `RGI.__call__`

What needs to be added to the document for `RGI.__call__`?

**Comment by ev-br** [comment]
> What needs to be added to the document for RGI.__call__?

I did not manage to figure the shape of the result and it's relationship to values.shape and xi.shape without looking at interpn docs --- which does list it explicitly.

**Comment by j-bowhay** [comment]
@ev-br think this is ready for review now, have updated the documentation and modified your test to test for more than 1 trailing dimension

**Comment by j-bowhay** [comment]
Will rewrite the pr description to something more representative of what this pr now is when I get some time over the next few days

File: scipy/interpolate/tests/test_rgi.py
Comment (ev-br) [inline scipy/interpolate/tests/test_rgi.py:509]: Nit: would be nice to add `atol=1e-14` --- this checks for equality in the floating-point, so let's make it more explicit. Only worth doing if there are other changed to lump it to.

File: scipy/interpolate/tests/test_rgi.py
Comment (ev-br) [inline scipy/interpolate/tests/test_rgi.py:512]: Oh wow. Might be an overkill, I *think* it'd be enough to only test with `num_trailing_dims = 2` or 3 as a single test. See also below.

File: scipy/interpolate/tests/test_rgi.py
Comment (ev-br) [inline scipy/interpolate/tests/test_rgi.py:541]: Aha. I'm going to be annoying and ask to restructure this test. Consider num_trailing_dims=2. Let `values.shape = (6, 7, 8, 9, 2, 3)`, just for the sake of an example. The first four  entries are interpolation axes, and there are 2*3=6 trailing values to broadcast over. I think the test should actually loop over these six values, not only three values it does now. 

The shape of the result is checked above (line 533), so here we only check the values themselves. Options: 1) restrict to the test to two trailing dimensions, type the double loop manually; 2) reshape values to squash the trailing dimensions into one, keep the single for loop, restore the shape of the result before `assert_allclose` in line 548 below --- whichever feels more natural to you.

File: scipy/interpolate/tests/test_rgi.py
Comment (ev-br) [inline scipy/interpolate/tests/test_rgi.py:545]: not sure what this does and why, can you add a comment please

File: scipy/interpolate/tests/test_rgi.py
Comment (ev-br) [inline scipy/interpolate/tests/test_rgi.py:807]: Not sure if spline methods allow these "duck-typed values", and if we really need to worry about those. Can leave for a future enhancement if there is a user request.

File: scipy/interpolate/tests/test_rgi.py
Comment (ev-br) [inline scipy/interpolate/tests/test_rgi.py:606]: Also atol=1e-14 or some such, here and elsewhere.

**Review by ev-br — CHANGES_REQUESTED** [review]
Looks almost there! The main comment is to restructure the trailing dimensions test a bit, so that it loops over all values of the trailing dims, not only the last one. 

Whether to parametrize over the number of trailing dims... not sure it's worth it really. It either works with ndim > 1 or it does not, so a single ndim value look enough to me. Is there a deeper reason you want it that way?

File: scipy/interpolate/_rgi.py
Comment (ev-br) [inline scipy/interpolate/_rgi.py:278]: Was I saying already that this PR is unexpectedly effective in smoking out pre-existing issues?

Looking at the new test_nonscalar_values_2 made me realize this statement is strictly speaking incorrect? Ditto the matching statement in `interpn`, where this line come from.
Having `xi.shape == (4,)`  (a 4D interpolation) and `values.shape=(6, 7, 8, 9, 2, 3)` returns an array of shape `(1, 2, 3)`. 
This only happens for 1D `xi` arrays, as test_nonscalar_values shows: `xi.shape=(7, 11, 4)` and `values.shape=(6, 6, 6, 6, 8)` leads to the result of shape `(7, 11, 8)`, which agrees to what's written in the docstring. 

This behavior is already observed on main for linear and nearest modes; this PR makes new modes consistent, so there's nothing to fix in this PR. Documentation might be nice to update though in a follow-up PR. Not sure how to phrase this succinctly, maybe numpy wizards could suggest the right wording.

File: scipy/interpolate/tests/test_rgi.py
Comment (ev-br) [inline scipy/interpolate/tests/test_rgi.py:523]: This makes the last dimension have length 4, equal to the `len(sample)`. Would be nice to help future readers of this to have all dimensions different. No need to hold this PR for it, will fix in a follow-up PR.

File: scipy/interpolate/tests/test_rgi.py
Comment (ev-br) [inline scipy/interpolate/tests/test_rgi.py:543]: Had to look up what expand_dims does, and this prompted the `xi.shape[:-1]` confusion above.

File: scipy/interpolate/tests/test_rgi.py
Comment (ev-br) [inline scipy/interpolate/tests/test_rgi.py:759]: `.T` is a bit confusing(-ly different from `.transpose(1, 2, 0)` in the previous test)

**Review by ev-br — APPROVED** [review]
Looks good to me now!

There are several minor issues, none of them is worth holding this PR for; let's merge and address those in a follow-up or several.

**Comment by ev-br** [comment]
The PR LGTM, the CI failure in linalg/test_decomp is unrelated. 27 commits is a bit too much here, needs some squashing. We typically prefer to not squash-merge multi-author PRs, but here it's OK so squash-merge it is. 

Merging, thanks a whole lot @j-bowhay!

This also needs a release note entry, or two in fact: one is that interpn now supports tensor-product interpolation methods ('slinear', 'cubic', 'quintic' and 'pchip'), and that these methods now broadcast across non-scalar `values` arrays.

File: scipy/interpolate/tests/test_rgi.py
Comment (ev-br) [reply scipy/interpolate/tests/test_rgi.py:523]: The fix is in https://github.com/scipy/scipy/pull/16829

File: scipy/interpolate/tests/test_rgi.py
Comment (j-bowhay) [reply scipy/interpolate/tests/test_rgi.py:800]: So I did look at adding the spline methods to this test but it required modifying the duck typed object to add things like ndims etc.

**Comment by j-bowhay** [comment]
Have updated the release notes. Thanks for your help getting this one over the line @ev-br

File: scipy/interpolate/tests/test_rgi.py
Comment (ev-br) [reply scipy/interpolate/tests/test_rgi.py:800]: Let's wait for a user request I'd say. This is a bit esoteric.

File: scipy/interpolate/_rgi.py
Comment (j-bowhay) [reply scipy/interpolate/_rgi.py:278]: So if I am understanding correctly the behaviour is:
if `xi.ndim == 1`:
`values_x : ndarray, 1 + values.shape[ndim:]`
if `xi.ndim > 1`:
`values_x : ndarray, shape xi.shape[:-1] + values.shape[ndim:]`

Comment (ev-br) [reply scipy/interpolate/_rgi.py:278]: Yes, that's my understanding as well (modulo a typo above: for 1D xi, the shape is `(1,) + values.shape[ndim:]` , with a one-element tuple.

Comment (j-bowhay) [reply scipy/interpolate/_rgi.py:278]: As far as I am aware the return type/shape needs to go on one line for it to render properly so I am not sure how this can neatly be documented?

Comment (ev-br) [reply scipy/interpolate/_rgi.py:278]: How about keeping this as is and adding a clarifying note in the Notes section

**Pull request merged**
